modules/images/image-loading-optimization/storage/data.php

```diff
@@ -0,0 +1,311 @@
+ *
+ * TODO: For non-singular requests, consider adding the post IDs from The Loop to ensure publishing a new post will invalidate the cache.
+ *
+ * @return array Normalized query vars.
+ */
+function ilo_get_normalized_query_vars() {
+	global $wp;
+
+	// Note that the order of this array is naturally normalized since it is
+	// assembled by iterating over public_query_vars.
+	$normalized_query_vars = $wp->query_vars;
+
+	// Normalize unbounded query vars.
+	if ( is_404() ) {
+		$normalized_query_vars = array(
+			'error' => 404,
+		);
+	}
+
+	return $normalized_query_vars;
+}
+
+/**
+ * Gets slug for page metrics.
+ *
+ * @see ilo_get_normalized_query_vars()
+ *
+ * @param array $query_vars Normalized query vars.
+ * @return string Slug.
+ */
+function ilo_get_page_metrics_slug( array $query_vars ) {
+	return md5( wp_json_encode( $query_vars ) );
+}
+
+/**
+ * Compute nonce for storing page metrics for a specific slug.
+ *
+ * This is used in the REST API to authenticate the storage of new page metrics from a given URL.
+ *
+ * @see wp_create_nonce()
+ * @see ilo_verify_page_metrics_storage_nonce()
+ *
+ * @param string $slug Page metrics slug.
+ * @return string Nonce.
+ */
+function ilo_get_page_metrics_storage_nonce( $slug ) {
+	return wp_create_nonce( "store_page_metrics:{$slug}" );
+}
+
+/**
+ * Verify nonce for storing page metrics for a specific slug.
+ *
+ * @see wp_verify_nonce()
+ * @see ilo_get_page_metrics_storage_nonce()
+ *
+ * @param string $nonce Page metrics storage nonce.
+ * @param string $slug  Page metrics slug.
+ * @return int|false 1 if the nonce is valid and generated between 0-12 hours ago,
+ *                   2 if the nonce is valid and generated between 12-24 hours ago.
+ *                   False if the nonce is invalid.
+ */
+function ilo_verify_page_metrics_storage_nonce( $nonce, $slug ) {
+	return wp_verify_nonce( $nonce, "store_page_metrics:{$slug}" );
+}
+
+/**
+ * Unshift a new page metric onto an array of page metrics.
+ *
+ * @param array $page_metrics          Page metrics.
+ * @param array $validated_page_metric Validated page metric. See JSON Schema defined in ilo_register_endpoint().
+ * @return array Updated page metrics.
+ */
+function ilo_unshift_page_metrics( array $page_metrics, array $validated_page_metric ) {
+	array_unshift( $page_metrics, $validated_page_metric );
+	$breakpoints          = ilo_get_breakpoint_max_widths();
+	$sample_size          = ilo_get_page_metrics_breakpoint_sample_size();
+	$grouped_page_metrics = ilo_group_page_metrics_by_breakpoint( $page_metrics, $breakpoints );
+
+	foreach ( $grouped_page_metrics as &$breakpoint_page_metrics ) {
+		if ( count( $breakpoint_page_metrics ) > $sample_size ) {
+
+			// Sort page metrics in descending order by timestamp.
+			usort(
+				$breakpoint_page_metrics,
+				static function ( $a, $b ) {
+					if ( ! isset( $a['timestamp'] ) || ! isset( $b['timestamp'] ) ) {
+						return 0;
+					}
+					return $b['timestamp'] <=> $a['timestamp'];
```

Comment:
nice!

Reply:
First time I've ever gotten to use this 😄

Reply:
i had to look it up to remember what it does :)